kaitaistruct.py

```diff
@@ -335,30 +349,52 @@ def bytes_terminate(data, term, include_term):
 
     @staticmethod
     def process_xor_one(data, key):
-        if PY2:
-            return bytes(bytearray(v ^ key for v in bytearray(data)))
-        else:
-            return bytes(v ^ key for v in data)
+        if key == 0:
+            return data
+
+        return integers2bytes(v ^ key for v in bytes2integers(data))
 
     @staticmethod
     def process_xor_many(data, key):
-        if PY2:
-            return bytes(bytearray(a ^ b for a, b in zip(bytearray(data), itertools.cycle(bytearray(key)))))
-        else:
-            return bytes(a ^ b for a, b in zip(data, itertools.cycle(key)))
+        if len(key) == 1:
+            return KaitaiStream.process_xor_one(data, ord(key))
+        if len(key) <= 64 and key == b'\x00' * len(key):
```

Comment:
why 64? Why not, for exaple a quarter of message length

Reply:
Its to make it O(1) time, in case both key and data was megabyte sized. If so then it skips the check and just does the xoring.

Reply:
>Its to make it O(1) time, in case both key and data was megabyte sized. If so then it skips the check and just does the xoring.

Again, IMHO it is waste of  resources, I assume xoring to zero string as improbable, in most cases it is waste of resources.

let we have the probability of the zero array is p, N is the bit-length of data, K is the bit-length of key.

the complexity in the case of zero is
O(K)
(in fact `2*K` because we create an array and then compare it)

otherwise

`O(K+N)`

So the average complexity of zero-checking case (we check key for zero) is

`c=K+(1-p)*N`

And the average complexity of non-checking case is always `N`.

So we need to select by the clause
`K+(1-p)*N<N`

`K<p*N`

Now let's model p.
If each bit of K is randomly chosen from {1, 0} then 
`p = 2^-K`

Reply:
Please read the reply below and withdraw this one review.